### PR TITLE
feat: voice-chat plugin

### LINKED ISSUE #2375: [Bounty] Clubhouse on top of Twitter
## Catalogue

- Platform
  - [x] Gitcoin
  - [ ] Hackerlink
  - [ ] Mask
- Experience
  - [ ] Beginner
  - [ ] Intermediate
  - [x] Advanced
- Category
  - [ ] Mask Core Feature
  - [x] Mask Plugin
  - [ ] Security/Cryptography Feature
  - [ ] Documentation
  - [ ] Design
  - [ ] Other
    - [ ] /_List it here_/
- Time Commitment
  - [x] 4 weeks
- Bounty Size
  - [x] 2500 USD

## Part 1: Problem Description
Clubhouse has gone viral on the Internet the past week, everywhere in the world. It shows the potential of a socialized audio-only chat room. The success of the Clubhouse is not simply relying on its core audio chat feature. It consists of quite a lot of tweaks and details on their design. Therefore, this bounty is not to copy it. Instead, we just want to show the potential of Mask, being a middle layer on the existing Internet, that can integrate an audio chat room directly on top of Twitter with minimal efforts.

The idea is to inject a chat room in a single tweet, where each person who sees it and have Mask installed can choose to join the room to listen to the discussion, or, raise their hand to speak. In another word, each tweet is a unique chat room, which is designed to be one, with the audio chat functionality. This plugin basically turns each RESTFUL tweet into a REAL TIME discussion. On top of a social media with enough social feature plus the access control system provided by Mask, I believe this would bring our users an alternative choice on the existing social media and it should also bring our (potential) developers more creativeness on how we can play with Mask.

## Part 2: Resources
https://www.joinclubhouse.com/
https://github.com/bestony/neshouse -> an open-source implementation by @bestony in 55 hours

## Part 3: Contact and Support
@yisiliu

### PR DESCRIPTION
**Pull request by septs**
closes #2375

### CODE REVIEW

File: packages/maskbook/src/plugins/Voicechat/define.tsx [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import type { VoiceChatMetadata } from './types'
+import { VoicechatDialog } from './UI/VoicechatDialog'
+import { VoicechatInlay } from './UI/VoicechatInlay'
+import { VoicechatMetadataReader } from './utils'
+
+const [VoicechatCompositionEntry, VoicechatDialogUI] = createCompositionDialog('🔊 Voicechat', (props) => (
+    <VoicechatDialog open={props.open} onConfirm={props.onClose} onDecline={props.onClose} />
+))
+
+export const VoicechatPluginDefine: PluginConfig = {
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Voicechat/define.tsx:13]: Please use the new infra to register the plugin

**Review by Jack-Works — CHANGES_REQUESTED** [review]

**Comment by septs** [comment]
@nykma plz deploy https://github.com/DimensionDev/voicechat-server service

File: packages/maskbook/src/plugins/Voicechat/SNSAdaptor/VoicechatInlay.tsx [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+export function VoicechatInlay(props: VoicechatInlayProps) {
+    const postIdentifier = usePostInfoDetails.postIdentifier()
+    const postBy = usePostInfoDetails.postBy()
+
+    const [peerMediaElements, setPeerMediaElements] = useState<Record<string, PeerMediaElement>>()
+    const [localPeerMediaElement, setLocalPeerMediaElement] = useState<PeerMediaElement | null>()
+    const [mediaDevices, setMediaDevices] = useState<MediaDeviceInfo[]>([])
+    const [activeMediaDevice, setActiveMediaDevice] = useState<string>('')
+    const [isMuted, setIsMuted] = useState<boolean>(false)
+    const [isJoining, setIsJoining] = useState<boolean>(false)
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Voicechat/SNSAdaptor/VoicechatInlay.tsx:187]:
```suggestion
    const [activeMediaDevice, setActiveMediaDevice] = useState('')
    const [isMuted, setIsMuted] = useState(false)
    const [isJoining, setIsJoining] = useState(false)
```

File: packages/maskbook/src/plugins/Voicechat/SNSAdaptor/index.tsx [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        dialog({ open, onClose }) {
+            return <VoicechatDialog open={open} onConfirm={onClose} onDecline={onClose} />
+        },
+    },
+}
+
+export default definition
+
+function isActive(payload: VoiceChatMetadata) {
+    return !!payload ? 'Voicechat enabled' : 'Voicechat disabled'
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Voicechat/SNSAdaptor/index.tsx:29]: `VoiceChat`?

File: packages/maskbook/src/plugins/Voicechat/utils/voicechat.ts [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+                join_chat_channel(id, { username })
+            },
+            () => {},
+            username,
+        )
+    })
+
+    setTimeout(() => {
+        if (!signalingSocket?.connected) {
+            alert('Could not connect to voice server')
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Voicechat/utils/voicechat.ts:56]: Do we really need to popup an alert here?

Comment (septs) [reply packages/maskbook/src/plugins/Voicechat/utils/voicechat.ts:56]: i don't known

File: packages/maskbook/src/plugins/Voicechat/constants.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+export const VoiceChatPluginID = 'com.maskbook.voicechat'
+export const VOICECHAT_META_KEY_1 = 'com.maskbook.voicechat:1'
+
+export const SIGNALING_SERVER = 'http://localhost:3000'
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Voicechat/constants.ts:4]: 👀

File: packages/maskbook/src/plugins/Voicechat/SNSAdaptor/index.tsx [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        const metadata = VoicechatMetadataReader(props.message.meta)
+        if (!metadata.ok) return null
+        return <VoicechatInlay metadata={metadata.unwrap()} />
+    },
+    CompositionDialogMetadataBadgeRender: new Map([[VOICECHAT_META_KEY_1, isActive]]),
+    CompositionDialogEntry: {
+        label: VoicechatDialog,
+        dialog({ open, onClose }) {
+            return <VoicechatDialog open={open} onConfirm={onClose} onDecline={onClose} />
+        },
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Voicechat/SNSAdaptor/index.tsx:22]: I suggest changing this to 
```suggestion
        dialog: VoicechatDialog,
```

And change props of `VoicechatDialog` to fit this

File: packages/maskbook/src/plugins/Voicechat/SNSAdaptor/VoicechatDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+                        />
+                    }
+                    label={t('plugin_voicechat_activate_switch')}
+                />
+                <Box marginTop={1}>
+                    <TextField
+                        label={t('plugin_voicechat_custom_server_url')}
+                        value={customServerURL}
+                        onChange={(e) => setCustomServerURL(e.target.value)}
+                        fullWidth={true}
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Voicechat/SNSAdaptor/VoicechatDialog.tsx:83]:
```suggestion
                        fullWidth
```

**Comment by septs** [comment]
@DimensionDev/qa ready for test

File: packages/maskbook/src/_locales/en/messages.json [hunk 8]
```diff
@@ -610,6 +609,7 @@
     "plugin_ito_unlocking_symbol": "Unlocking {{symbol}}",
     "plugin_ito_continue": "Continue",
     "plugin_ito_view_on_explorer": "View on Explorer",
+    "plugin_ito_placeholder_when_token_unselected": "Please Select a Token first",
```
Comment (guanbinrui) [inline packages/maskbook/src/_locales/en/messages.json:612]:
```suggestion
    "plugin_ito_placeholder_when_token_unselected": "Please select a token first",
```

**Review by guanbinrui — DISMISSED** [review]

File: packages/maskbook/src/locales/ko-KR.json [hunk 9]
```diff
@@ -508,6 +508,10 @@
     "plugin_ito_amount_unlocked": "{{amount}} {{symbol}} 언락됨",
     "plugin_ito_amount_unlocked_infinity": "언락된 인피니티 {{symbol}}",
     "plugin_ito_unlocking_symbol": "{{symbol}} 언락 중",
+    "plugin_ito_continue": "",
```
Comment (Jack-Works) [inline packages/maskbook/src/locales/ko-KR.json:511]: Don't add those empty string because we're integrating with Crowdin, it don't like empty strings.

**Pull request closed without merging**